_posts/2020-02-13-spring-boot-application-events-explained.md

```diff
@@ -0,0 +1,181 @@
+
+## Throwing an Application Event
+
+Application events are sent by using the Spring Framework’s event publishing mechanism. Prior to Spring 4.2, it was required to extend `ApplicationEvent` for latter it is no longer required.
+
+In the below example, we will create a custom event called `UserCreationEvent` to store user-related data.
+
+```java
+public class UserCreatedEvent extends ApplicationEvent {
+	private String name;
+
+	public UserCreatedEvent(String name) {
+		super(name);
+		this.name = name;
+	}
+
+	public String getName() {
+		return this.name;
+	}
+}
+```
+
+## Publishing an Application Event
+
+Spring gives the flexibility to publish any arbitrary event and not force you to extend from `ApplicationEvent`. The `ApplicationEventPublisher` interface has been extended to allow you to publish any object. When this object is not an `ApplicationEvent`, Spring will wrap it in a `PayloadApplicationEvent` for you.
+
+Now that we have `UserCreatedEvent`. In order to publish it to its respective listeners, we need `ApplicationEventPublisher`.
+
+```java
+@Component
+public class Publisher {
+	@Autowired
+	private ApplicationEventPublisher publisher;
+
+	public void publishEvent(final String name) {
+		publisher.publishEvent(new UserCreatedEvent(name));
+	}
+}
+```
+
+## Listening to an Application Event
+
+Now that you know how to create and publish a custom event, let us see how to start listening to the event. An event can have multiple listeners doing different work based on the application requirement. By default spring events are synchronous, it blocks until all listeners finish processing the event.
+
+There are two ways to define a listener. One with `@EventListener` annotation and another by implementing `ApplicationListener`.
+
+Note: Listener has to be a bean managed by spring.
+
+### Annotation-Driven
+
+From Spring 4.1 it is now possible to simply annotate a method of a managed bean with `@EventListener` to automatically register an `ApplicationListener` matching the signature of the method. No additional configuration is necessary with annotation-driven configuration enabled.
+
+```java
+@EventListener
+public void handleUserCreatedEvent(UserCreatedEvent event) {
+	System.out.println("User created event triggered. Created user: " + event.getName());
+}
+```
+
+For the methods annotated with `@EventListener` and defined as a non-void return type, Spring will send that result as a new event for you.
+
+Spring allows your listener to be triggered by a particular scenario using a condition attribute. The event will be handled if the expression evaluates to boolean true or one of the following strings: "true", "on", "yes", or "1". Method arguments are exposed via their names. The condition expression also exposes a “root” variable with raw `ApplicationEvent (#root.event)` and the actual method arguments `(#root.args)`
+
+### Implementing ApplicationListener
+
```

Comment:
Yes. I have added explanations wherever it was required now. Hope it looks better now

Reply:
Have a look at my other comments. You introduced explanations, but most of the time, before the code example. The reader needs to see the code example first, and only then the explanation, otherwise they will have to skip forward to understand.